### PR TITLE
Add `#[track_caller]` to lint related diagnostic functions

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This fixes locations reported by `-Ztrack-diagnostics`.

### CODE REVIEW

**Review by compiler-errors — APPROVED** [review]
r=me

File: compiler/rustc_lint/src/levels.rs [hunk 1]
```diff
@@ -1028,6 +1030,7 @@ impl<'s, P: LintLevelsProvider> LintLevelsBuilder<'s, P> {
         struct_lint_level(self.sess, lint, level, src, span, msg, decorate)
     }
 
+    #[track_caller]
     pub fn emit_spanned_lint(
```
Comment (WaffleLapkin) [inline compiler/rustc_lint/src/levels.rs:1034]: Should this and similar have `#[rustc_lint_diagnostics]` also? it seems like there are a lot of lint-related diagnostic functions without `#[rustc_lint_diagnostics]`...

Comment (compiler-errors) [reply compiler/rustc_lint/src/levels.rs:1034]: It shouldn't matter, I don't even know if they trigger `-Ztreat-err-as-bug` which is presumably one of the main usages of tracking caller.

Comment (compiler-errors) [reply compiler/rustc_lint/src/levels.rs:1034]: edit: im dumb

`rustc_lint_diagnostics` shouldn't matter, ideally you'd add that to all the fns that we should deny outside of translatable diagnostics derives but that should be in a separate pr.

**Pull request merged**